keps/sig-network/0006-20180314-coredns-GA-proposal.md

```diff
@@ -0,0 +1,132 @@
+    * Making an alias for an external name [#39792](https://github.com/kubernetes/kubernetes/issues/39792)
+    * Dynamically adding services to another domain, without running another server [#55](https://github.com/kubernetes/dns/issues/55)
+    * Adding an arbitrary entry inside the cluster domain (for example TXT entries [#38](https://github.com/kubernetes/dns/issues/38))
+    * Verified pod DNS entries (ensure pod exists in specified namespace)
+    * Limit PTR replies to the specific CIDRs [#125](https://github.com/kubernetes/dns/issues/125)
+    * Serve DNS for selected namespaces [#132](https://github.com/kubernetes/dns/issues/132)
+    * Serve DNS based on a label selector.
+    * Support for wildcard queries (e.g., `*.namespace.svc.cluster.local` returns all services in `namespace`)
+    
+* By default, the user experience would be unchanged. For more advanced uses, existing users would need to modify the ConfigMap that contains the CoreDNS configuration file.
+* Since CoreDNS has more supporting features than kube-dns, there will be no path to retain the CoreDNS configuration in case a user wants to switch to kube-dns.
+
+#### Configuring CoreDNS
+
+The CoreDNS configuration file is called a `Corefile` and syntactically is the same as a [Caddyfile](https://caddyserver.com/docs/caddyfile). The file consists of multiple stanzas called _server blocks_.
+Each of these represents a set of zones for which that server block should respond, along with the list of plugins to apply to a given request. More details on this can be found in the 
+[Corefile Explained](https://coredns.io/2017/07/23/corefile-explained/) and [How Queries Are Processed](https://coredns.io/2017/06/08/how-queries-are-processed-in-coredns/) blog entries.
+
+The following can be expected when CoreDNS is graduated to GA.
+
+#### Kubeadm
+
+* The CoreDNS feature-gates flag will be set to default and marked as GA.
+* On a fresh install of Kubernetes via kubeadm by specifying the `kubeadm init` command, CoreDNS will be the default installed DNS add-on.
+* In case a user wants to install kube-dns instead of CoreDNS, they have to set the feature-gate of CoreDNS to false. `--feature-gates=CoreDNS=false`
+* For users upgrading Kubernetes via kubeadm, it will install CoreDNS whether the user had kube-dns or CoreDNS in a previous kubernetes version (Unless the CoreDNS feature-gate is disabled).
+* The configmap of a previously installed kube-dns will be automatically translated to the equivalent CoreDNS configmap.
+
+#### Kube-up
+
+* CoreDNS set as the default DNS service in place of kube-dns.
+
+#### Minikube
+
+* CoreDNS to be enabled by default in the add-on manager, with kube-dns disabled by default.
+ 
+## Unresolved questions
+
+* Fate of kube-dns in future releases, i.e. deprecation path.
+
+## Graduation Criteria
+
+* Passing all e2e conformance and DNS related tests.
+
```

Comment:
Scalability tests

Reply:
@bowei : what have you in mind ?
- a manual scale test, as we did here : https://github.com/kubernetes/community/pull/1100#issuecomment-337747482 (using Kubemark)
- or an extension of these [perf-test for DNS ](https://github.com/kubernetes/perf-tests)(adapting those for CoreDNS) ? 
- or another kind of scalability included in kubernetes/tests/e2e ?

Reply:
Sorry about the terseness, I mean that we should get the CoreDNS running as a part of regular k8s scale tests.

Having it possible to be run as target of perf-test for DNS would be great.

Reply:
We have looked at the perf-test for DNS, we can do that but it will require a lot of changes to those tests. They are geared toward discovering/testing varying levels of resources devoted to dnsmasq and kube-dns.

Reply:
I think we should have an integration style test for CoreDNS in that style given the surprises we've experienced wrt to resource limits etc from the past.

Reply:
@bowei : Let me verify I understand your 2 points:

### CoreDNS as part of the regular k8s scale runs (e.g. up to 5k nodes)

That means have CoreDNS be part of the regular k8s scale runs at 5k nodes
Therefore ensure we have a DNS e2e tests with the flag "[Feature:Performance]". 
It will automatically run with all perf-tests like "ci-kubernetes-kubemark-gce-scale" (which is the one running 5k nodes with Kubemark)


###  integration test style resource determination

From this sentence, and your comment before, I understand you advise to extend those **[perf-test for DNS]( https://github.com/kubernetes/perf-tests/tree/master/dns)** and create some dedicated tests for CoreDNS to measure the resources used by CoreDNS under stress of queries.

**is my understanding correct ?**

Reply:
lgtm